popovers: Hide left sidebar actions popover for empty topics.

saubhagya-patel opened the pull request:
This PR changes the behavior of the topic actions popover (displayed from the left sidebar via the `...` menu) to no longer appear for empty topics (topics with no messages).
Instead, in place of the popover a message is displayed: "There are no messages in this topic.", along with the name of the topic (styled like the deactivated user line).

Fixes: #32098.
[CZO thread](https://chat.zulip.org/#narrow/channel/9-issues/topic/Error.20moving.20topic.20with.20no.20messages/near/1966728)


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
**Before Changes**
<details>
<summary>GIF</summary>

![empty before](https://github.com/user-attachments/assets/cfd351f0-e552-4de3-a943-312c43de928b)

</details>

<details>
<summary>Image</summary>

![Screenshot 2024-11-04 223613](https://github.com/user-attachments/assets/773f6e02-3636-4961-b9ee-b192d050bce7)


</details>

**After Changes**
<details>
<summary>GIF</summary>

![empty after](https://github.com/user-attachments/assets/33381a10-cf60-4eab-af9a-30e53e221898)

</details>

<details>
<summary>Image</summary>

![Screenshot 2024-11-04 222900](https://github.com/user-attachments/assets/cd143bc0-4214-473f-89cb-60c39b4e5470)

</details>

**Please note that the left sidebar actions popover for non-empty topics remains unchanged.**

<details>
<summary>GIF: Non-empty topic actions popover.</summary>

![non empty](https://github.com/user-attachments/assets/4e097520-51f5-4c0d-9010-fd819ccc281e)

</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
Thanks @saubhagya-patel!

@sayamsamal, would you be up for reviewing this one?

sayamsamal commented on web/src/popover_menus_data.ts:
```diff
@@ -11,6 +11,7 @@ import * as hash_util from "./hash_util";
 import {$t} from "./i18n";
 import * as message_edit from "./message_edit";
 import * as message_lists from "./message_lists";
+import {get_messages_in_topic} from "./message_util";
```
Please use [namespace imports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#namespace_import), instead of [named imports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#named_import) — it helps with the code readability.

saubhagya-patel replied:
I apologize for the inconvenience caused and will remember this in the future. Thank you.

sayamsamal commented on web/src/topic_popover.js:
```diff
@@ -63,6 +63,10 @@ export function initialize() {
                     return;
                 }
 
+                if (instance.context.is_topic_empty) {
+                    return;
+                }
+
```
Was the need for this guard statement due to the ClipboardJS initialization below?

saubhagya-patel commented:
I have addressed the issue of named import as mentioned in [this comment](https://github.com/zulip/zulip/pull/32234#discussion_r1838055919).

>  There may still be some cases where we are unable to confirm with accuracy if the topic is empty, but encapsulating those cases would require server calls which would be expensive?

As you mentioned, I think I will wait for the insights. However, if I find any useful method to confirm whether the topic is empty, I will update this PR. Thank you for the valuable feedback.

sayamsamal commented:
I feel @timabbott would be better equipped to answer that, so marking it for integration.

timabbott commented on web/src/popover_menus_data.ts:
```diff
@@ -248,13 +250,15 @@ export function get_topic_popover_content_context({
     const visibility_policy = user_topics.get_topic_visibility_policy(sub.stream_id, topic_name);
     const all_visibility_policies = user_topics.all_visibility_policies;
     const is_spectator = page_params.is_spectator;
+    const is_topic_empty = message_util.get_messages_in_topic(stream_id, topic_name).length === 0;
```
This check is not correct unless we have all the messages in the local cache. I think the correct check is a bit subtle.

I think we want to call the thing that we'll compute `is_topic_definitely_empty`, and I think its value can be true if and only if all of the following are true:
- It's the currently active channel/topic pair (`narrow_state` should support checking this)
- `message_lists.current.data.empty()` is true
- `message_lists.current.data.fetch_status.has_found_oldest()` is true

(We only show empty topics in the sidebar if it's the current one)

saubhagya-patel replied:
Thank you for the valuable feedback. I will work on this.

timabbott commented on web/src/topic_popover.js:
```diff
@@ -63,6 +63,10 @@ export function initialize() {
+
                 $popper.on("change", "input[name='sidebar-topic-visibility-select']", (e) => {
                     const start_time = Date.now();
                     const visibility_policy = Number.parseInt(
```
It might be a bug to not show the topic visibility feature here; that should be able to operate correctly on the current topic, even if it's empty.

saubhagya-patel commented:
I have updated this PR following the review received. The following changes were made:

- Introduced an `is_topic_definitely_empty` function which checks if the current topic is empty based on the review received in [this](https://github.com/zulip/zulip/pull/32234#discussion_r1851183766) comment.
- As stated in [this](https://github.com/zulip/zulip/pull/32234#discussion_r1851184336) comment, the left sidebar topic actions popover now also shows the visibility actions for empty topics. I have attached updated screenshots.

  <details>
  <summary>Screenshots of the updated popover.</summary>
  
   ![Screenshot 2024-11-26 113542](https://github.com/user-attachments/assets/84c37f4f-dda5-432b-99af-a89479934eb4)
   
   ![Screenshot 2024-11-26 113635](https://github.com/user-attachments/assets/55d8b9c2-7671-47b6-94f3-f50e688889f9)
  
  </details>

- No changes have been made to the popover for non-empty topics.

Please let me know if any further adjustments are needed. I apologize for the delay in updates as my college exams are ongoing. Thank you.

saubhagya-patel commented:
@timabbott can you please take a look at the updated changes when you have a moment. Thank you.

timabbott commented:
Looks great, merged, thanks @saubhagya-patel!

Actually I did miss one thing, `definitely` was misspelled; fixed in 7f7698e9b9d3431b984d83f05f89b2b97e0b8634.
